tests/lms/views/decorators/h_api_test.py

```diff
@@ -0,0 +1,310 @@
+            "users",
+            {
+                "username": "test_username",
+                "display_name": "test_display_name",
+                "authority": "TEST_AUTHORITY",
+                "identities": [
+                    {
+                        "provider": "test_provider",
+                        "provider_unique_id": "test_provider_unique_id",
+                    }
+                ],
+            },
+        )
+
+    def test_it_continues_to_the_wrapped_function(
+        self, upsert_h_user, context, pyramid_request, wrapped
+    ):
+        returned = upsert_h_user(context, pyramid_request)
+
+        wrapped.assert_called_once_with(context, pyramid_request)
+        assert returned == wrapped.return_value
+
+    @pytest.fixture
+    def upsert_h_user(self, wrapped):
+        # Return the actual wrapper function so that tests can call it directly.
+        return h_api.upsert_h_user(wrapped)
+
+
+@pytest.mark.usefixtures("hapi_svc")
+class TestCreateCourseGroup:
+    def test_it_does_nothing_if_the_feature_isnt_enabled(
+        self, upsert_course_group, context, pyramid_request, wrapped, hapi_svc
+    ):
+        # If the auto provisioning feature isn't enabled for this application
+        # instance then upsert_course_group() doesn't do anything - just calls the
+        # wrapped view.
+        context.provisioning_enabled = False
+
+        returned = upsert_course_group(context, pyramid_request)
+
+        hapi_svc.patch.assert_not_called()
+        hapi_svc.put.assert_not_called()
+        wrapped.assert_called_once_with(context, pyramid_request)
+        assert returned == wrapped.return_value
+
+    def test_it_calls_the_group_update_api(
+        self, upsert_course_group, context, pyramid_request, hapi_svc
+    ):
+        upsert_course_group(context, pyramid_request)
+
+        hapi_svc.patch.assert_called_once_with(
+            "groups/test_groupid", {"name": "test_group_name"}
+        )
+
+    def test_it_raises_if_updating_the_group_fails(
+        self, upsert_course_group, context, pyramid_request, hapi_svc
+    ):
+        # If the group update API call fails for any non-404 reason then the
+        # view raises an exception and an error page is shown.
+        hapi_svc.patch.side_effect = HAPIError("Oops")
+
+        with pytest.raises(HAPIError, match="Oops"):
+            upsert_course_group(context, pyramid_request)
+
+    def test_if_the_group_doesnt_exist_it_creates_it(
+        self, upsert_course_group, context, pyramid_request, hapi_svc
+    ):
+        hapi_svc.patch.side_effect = HAPINotFoundError()
+
+        upsert_course_group(context, pyramid_request)
+
+        hapi_svc.put.assert_called_once_with(
+            "groups/test_groupid",
+            {"groupid": "test_groupid", "name": "test_group_name"},
+            "acct:test_username@TEST_AUTHORITY",
+        )
+
+    def test_if_the_group_doesnt_exist_and_the_user_isnt_allowed_to_create_groups_it_400s(
```

Comment:
This test name is awkward as it doesn't follow the `test_it_does_this_if_that` template...

perhaps `test_it_400s_if_group_doesnt_exist_and_user_isnt_allowed_to_create_groups` ?

Reply:
These kinda make me miss Mocha's `context()` thing, where in JS tests you can do:

```javascript
  describe('#indexOf()', function() {
    context('when not present', function() {
      it('should not throw an error', function() {
        ...
```

I think the pytest equivalent might be to use a fixture, and I might start experimenting with something like this:

```python
@pytest.mark.usefixtures("group_doesnt_exist", "user_isnt_allowed_to_create_groups")
def test_it_400s(...):
    ...
```

But we'd have to do something about the test name in pytest output (which I think might be doable).